### PR TITLE
dialects: Add a stencil.ApplyOp to gpu.LaunchOp lowering.

### PR DESCRIPTION
**Pull request by PapyChacal**
This is trial-and-error code, it needs a lot of cleanup.

This works on my machine :yum:

### CODE REVIEW

File: xdsl/transforms/experimental/ConvertStencilToLLMLIR.py [hunk 1]
```diff
@@ -110,32 +119,120 @@ def match_and_rewrite(self, op: LoadOp, rewriter: PatternRewriter, /):
...
-        entry = op.region.blocks[0]
+        body = prepare_apply_body(op, rewriter)
+        dim = len(op.lb.array.data)
 
-        for arg in entry.args:
-            arg_uses = set(arg.uses)
-            for use in arg_uses:
-                use.operation.replace_operand(use.index, op.args[use.index])
-            entry.erase_arg(arg)
+        threads_per_dim = [[1024, 32, 8][dim - 1]] * dim + [1] * (3 - dim)
```
Comment (georgebisbas) [inline xdsl/transforms/experimental/ConvertStencilToLLMLIR.py:154]: Try adding a comment to explain these options

Comment (superlopuh) [inline xdsl/transforms/experimental/ConvertStencilToLLMLIR.py:154]: 🤯

File: xdsl/transforms/experimental/ConvertStencilToLLMLIR.py [hunk 2]
```diff
@@ -110,32 +119,120 @@ def match_and_rewrite(self, op: LoadOp, rewriter: PatternRewriter, /):
...
+        ])
+
+        body.insert_block(index_compute_block, 0)
+        body.add_block(else_block)
+        launch.body.blocks[1].add_op(gpu.TerminatorOp.get())
+
+        # Replace with the loop and necessary constants.
+        rewriter.insert_op_before_matched_op(
+            [*cst_tpd, *bounds_cst, *blocks_divs, one, launch])
+        rewriter.erase_matched_op(False)
```
Comment (superlopuh) [inline xdsl/transforms/experimental/ConvertStencilToLLMLIR.py:224]:
```suggestion
        rewriter.erase_matched_op(safe_erase=False)
```
I had to look up what `False` did, we might want to force it to be a named parameter

**Review by superlopuh — APPROVED** [review]

File: xdsl/transforms/experimental/ConvertStencilToLLMLIR.py [hunk 3]
```diff
@@ -110,32 +117,137 @@ def match_and_rewrite(self, op: LoadOp, rewriter: PatternRewriter, /):
...
+            for bc in bounds_cst
+        ]
+
+        # We check that the indices are inbound
+        cmpis: list[arith.Cmpi | arith.AndI] = [
+            arith.Cmpi.from_mnemonic(ti, bc, "ult")
+            for ti, bc in zip(thread_int, bounds_int)
+        ]
+
+        # We "arrith.andi" all those inbound checks
```
Comment (webmiche) [inline xdsl/transforms/experimental/ConvertStencilToLLMLIR.py:212]:
```suggestion
        # We "arith.andi" all those inbound checks
```

**Review by webmiche — APPROVED** [review]
Nice, that's very cool!

Just some minor things.

**Pull request merged**